android-core/src/main/java/com/mparticle/internal/ConfigManager.java

```diff
@@ -212,25 +224,92 @@ static void deleteConfigManager(Context context) {
         }
     }
 
-    void saveConfigJson(JSONObject json) {
+    /**
+     * detrmine if the stored config age is greater than mMaxConfigAge and clear it from storage if it is
+     */
+    void checkConfigStaleness() {
+        Long storageDate = getConfigTimestamp();
+        if (storageDate == null) {
+            // migration step: if the current config does not have a timestamp, set one to the current time
+            setConfigTimestamp(System.currentTimeMillis());
+            storageDate = getConfigTimestamp();
+        }
+        if (mMaxConfigAge == null || mMaxConfigAge < 0) {
+            return;
+        }
+        if (mMaxConfigAge == 0) {
+            clearConfig();
+        } else {
+            if (System.currentTimeMillis() >= storageDate + TimeUnit.SECONDS.toMillis(mMaxConfigAge)) {
+                clearConfig();
+            }
+        }
+    }
+
+    String getConfig() {
+        return sPreferences.getString(CONFIG_JSON, null);
+    }
+
+    void setConfigTimestamp(Long timestamp) {
+        sPreferences.edit()
+                .putLong(CONFIG_JSON_TIMESTAMP, timestamp)
+                .apply();
+    }
+
+    @Nullable
+    Long getConfigTimestamp() {
+        if (sPreferences.contains(CONFIG_JSON_TIMESTAMP)) {
+            return sPreferences.getLong(CONFIG_JSON_TIMESTAMP, 0);
+        } else {
+            return null;
+        }
+    }
+
+    void saveConfigJson(JSONObject json, String etag, String lastModified) throws JSONException {
         if (json != null) {
-            sPreferences.edit().putString(CONFIG_JSON, json.toString()).apply();
+            Logger.info("Updating config to:\n" + json.toString());
+            sPreferences.edit()
+                    .putLong(CONFIG_JSON_TIMESTAMP, System.currentTimeMillis())
```

Comment:
noting that we could use the timestamp that the server returns with if we wanted, but I actually think using the device timestamp is better - in case the device timestamp is inaccurate at least the delta/comparison will be correct.

Reply:
agreed, I think it's pretty in line with the overall theme of this change; create some client-side/server-independent checks on config state